It's almost the same tty manager of gdm 2.20, except we don't support Sun VTs,
only Linux and FreeBSD.
Index: gdm-3.0.4/common/gdm-settings-keys.h
===================================================================
--- gdm-3.0.4.orig/common/gdm-settings-keys.h	2011-05-31 16:28:03.000000000 +0200
+++ gdm-3.0.4/common/gdm-settings-keys.h	2011-06-06 11:20:01.927523114 +0200
@@ -32,6 +32,7 @@ G_BEGIN_DECLS
 #define GDM_KEY_TIMED_LOGIN_ENABLE "daemon/TimedLoginEnable"
 #define GDM_KEY_TIMED_LOGIN_USER "daemon/TimedLogin"
 #define GDM_KEY_TIMED_LOGIN_DELAY "daemon/TimedLoginDelay"
+#define GDM_KEY_FIRST_VT "daemon/FirstVT"
 
 #define GDM_KEY_DEBUG "debug/Enable"
 
Index: gdm-3.0.4/daemon/gdm-server.c
===================================================================
--- gdm-3.0.4.orig/daemon/gdm-server.c	2011-05-31 16:52:04.000000000 +0200
+++ gdm-3.0.4/daemon/gdm-server.c	2011-06-06 11:20:33.539677601 +0200
@@ -48,6 +48,9 @@
 #include "gdm-common.h"
 #include "gdm-signal-handler.h"
 
+#include "gdm-settings-direct.h"
+#include "gdm-settings-keys.h"
+
 #include "gdm-server.h"
 
 extern char **environ;
@@ -673,6 +676,213 @@ gdm_server_spawn (GdmServer  *server,
         return ret;
 }
 
+#include <sys/ioctl.h>
+#include <sys/stat.h>
+#if defined (__linux__)
+#define GDM_USE_SYS_VT
+#define GDMCONSOLEDEVICE "/dev/tty0"
+#include <sys/vt.h>
+#elif defined (__FreeBSD__) || defined (__FreeBSD_kernel__)
+#define GDM_USE_CONSIO_VT
+#define GDMCONSOLEDEVICE "/dev/ttyv0"
+#include <sys/consio.h>
+
+static const char*
+__itovty (int val)
+{
+	static char str[8];
+	char* next = str + sizeof (str) - 1;
+
+	*next = '\0';
+	do {
+		*--next = "0123456789abcdefghigklmnopqrstuv"[val % 32];
+	} while (val /= 32);
+
+	return next;
+}
+#elif defined __GNU__
+#define GDM_USE_HURD_VT
+#else /* __linux__ */
+#error Platform not supported
+#endif
+
+
+#ifndef GDM_USE_HURD_VT
+static gchar *
+gdm_get_vt_device (int vtno)
+{
+   gchar *vtname = NULL;
+
+#if defined (GDM_USE_SYS_VT)
+     vtname = g_strdup_printf ("/dev/tty%d", vtno);
+#elif defined (GDM_USE_CONSIO_VT)
+     vtname = g_strdup_printf ("/dev/ttyv%s", __itovty (vtno - 1));
+#endif
+
+   return vtname;
+}
+
+static int
+open_vt (int vtno)
+{
+	char *vtname = NULL;
+	int fd = -1;
+
+	vtname = gdm_get_vt_device (vtno);
+
+	do {
+		errno = 0;
+		fd = open (vtname, O_RDWR
+#ifdef O_NOCTTY
+			   |O_NOCTTY
+#endif
+			   , 0);
+	} while G_UNLIKELY (errno == EINTR);
+
+	g_free (vtname);
+
+	return fd;
+}
+
+static int
+get_first_vt (void)
+{
+        int first_vt;
+        if (!gdm_settings_direct_get_int (GDM_KEY_FIRST_VT,
+                                          &first_vt))
+                first_vt = 1;
+        return first_vt;
+}
+#endif /* GDM_USE_HURD_VT */
+
+#if defined (GDM_USE_SYS_VT)
+
+static int
+get_free_vt (void)
+{
+        int fd, fdv;
+        int vtno;
+        unsigned short vtmask;
+        struct vt_stat vtstat;
+
+        do {
+                errno = 0;
+                fd = open (GDMCONSOLEDEVICE,
+                           O_WRONLY
+#ifdef O_NOCTTY
+                           |O_NOCTTY
+#endif
+                           , 0);
+        } while G_UNLIKELY (errno == EINTR);
+        if (fd < 0)
+                return -1;
+
+        if (ioctl (fd, VT_GETSTATE, &vtstat) < 0) {
+                VE_IGNORE_EINTR (close (fd));
+                return -1;
+        }
+
+        for (vtno = get_first_vt (), vtmask = 1 << vtno; vtstat.v_state & vtmask; vtno++, vtmask <<= 1);
+        if (!vtmask) {
+                VE_IGNORE_EINTR (close (fd));
+                return -1;
+        }
+
+        fdv = open_vt (vtno);
+        if (fdv < 0) {
+                VE_IGNORE_EINTR (close (fd));
+                return -1;
+        }
+        VE_IGNORE_EINTR (close (fdv));
+        VE_IGNORE_EINTR (close (fd));
+        return vtno;
+}
+
+#elif defined (GDM_USE_CONSIO_VT)
+
+static int
+get_free_vt (void)
+{
+        int fd, fdv;
+        int vtno;
+        GList *to_close_vts = NULL, *li;
+
+        do {
+                errno = 0;
+                fd = open (GDMCONSOLEDEVICE,
+                           O_WRONLY
+#ifdef O_NOCTTY
+                           |O_NOCTTY
+#endif
+                           , 0);
+        } while G_UNLIKELY (errno == EINTR);
+        if (fd < 0)
+                return -1;
+
+        if ((ioctl (fd, VT_OPENQRY, &vtno) < 0) || (vtno == -1)) {
+                VE_IGNORE_EINTR (close (fd));
+                return -1;
+        }
+
+        fdv = open_vt (vtno);
+        if (fdv < 0) {
+                VE_IGNORE_EINTR (close (fd));
+                return -1;
+        }
+
+        while (vtno < get_first_vt ()) {
+                int oldvt = vtno;
+                to_close_vts = g_list_prepend (to_close_vts,
+                                               GINT_TO_POINTER (fdv));
+
+                if (ioctl (fd, VT_OPENQRY, &vtno) == -1) {
+                        vtno = -1;
+                        break;
+                }
+
+                if (oldvt == vtno) {
+                        vtno = -1;
+                        break;
+                }
+
+                fdv = open_vt (vtno);
+                if (fdv < 0) {
+                        vtno = -1;
+                        break;
+                }
+        }
+
+        for (li = to_close_vts; li != NULL; li = li->next) {
+                VE_IGNORE_EINTR (close (GPOINTER_TO_INT (li->data)));
+        }
+        VE_IGNORE_EINTR (close (fdv));
+        VE_IGNORE_EINTR (close (fd));
+        return vtno;
+}
+
+#elif defined (GDM_USE_HURD_VT)
+
+static int
+get_free_vt (void)
+{
+        return -1;
+}
+
+#endif
+
+static char *
+gdm_get_empty_vt_argument (void)
+{
+        int vt = -1;
+
+        vt = get_free_vt ();
+
+        if (vt < 0)
+                return NULL;
+        else
+                return g_strdup_printf ("vt%d", vt);
+}
+
 /**
  * gdm_server_start:
  * @disp: Pointer to a GdmDisplay structure
@@ -684,10 +894,14 @@ gboolean
 gdm_server_start (GdmServer *server)
 {
         gboolean res;
+        char *vtarg;
+
+        vtarg = gdm_get_empty_vt_argument ();
 
         /* fork X server process */
-        res = gdm_server_spawn (server, NULL);
+        res = gdm_server_spawn (server, vtarg);
 
+        g_free (vtarg);
         return res;
 }
 
Index: gdm-3.0.4/data/gdm.schemas.in.in
===================================================================
--- gdm-3.0.4.orig/data/gdm.schemas.in.in	2011-05-31 16:28:03.000000000 +0200
+++ gdm-3.0.4/data/gdm.schemas.in.in	2011-06-06 11:20:01.931523138 +0200
@@ -47,6 +47,11 @@
       <signature>i</signature>
       <default>30</default>
     </schema>
+    <schema>
+      <key>daemon/FirstVT</key>
+      <signature>i</signature>
+      <default>7</default>
+    </schema>
 
     <schema>
       <key>debug/Enable</key>
